### PR TITLE
avoid flushing disk when doing state tree snapshots

### PR DESCRIPTION
**Pull request by whyrusleeping**
state tree snapshots are currently the number one thing in chain validation profiles, this should improve things a bit

### CODE REVIEW

**Comment by whyrusleeping** [comment]
tested vs latest testnet/3, on a chain ~300 blocks long, testnet/3 takes 40 seconds on lotus-bench import, while this branch takes 13.4

I also expect this difference to change dramatically as the chains get larger, as the time spent flushing increases over time.

File: chain/state/statetree.go [hunk 1]
```diff
@@ -26,16 +26,71 @@ type StateTree struct {
...
+			}
+
+			return &act.Act, nil
+		}
+	}
+	return nil, nil
+}
+
+func (ss *stateSnaps) setActor(addr address.Address, act *types.Actor) {
+	ss.layers[len(ss.layers)-1][addr] = streeOp{Act: *act}
```
Comment (magik6k) [inline chain/state/statetree.go:82]: Technically this can override actor that is marked for deletion, but I don't think this can happen in practice.

And setting a deleted actor should probably 'bring it back' anyways

File: chain/state/statetree.go [hunk 2]
```diff
@@ -26,16 +26,71 @@ type StateTree struct {
...
+
+func (ss *stateSnaps) mergeLastLayer() {
+	last := ss.layers[len(ss.layers)-1]
+	nextLast := ss.layers[len(ss.layers)-2]
+
+	for k, v := range last {
+		nextLast[k] = v
+	}
+
+	ss.dropLayer()
```
Comment (magik6k) [inline chain/state/statetree.go:64]: Probably warrants benchmarking - Deleting keys as we merge, then keeping the map instance could save us some allocations

File: chain/state/statetree.go [hunk 3]
```diff
@@ -26,16 +26,71 @@ type StateTree struct {
 	root  *hamt.Node
 	Store cbor.IpldStore
 
-	// Maps ID addresses to actors.
-	actorcache map[address.Address]*types.Actor
-	snapshots  []cid.Cid
+	// Maps ID addresses to actors, layered by snapshots
```
Comment (arajasek) [inline chain/state/statetree.go:29]: comment can be taken out

File: chain/state/statetree.go [hunk 4]
```diff
@@ -175,17 +236,13 @@ func (st *StateTree) Snapshot(ctx context.Context) error {
 	ctx, span := trace.StartSpan(ctx, "stateTree.SnapShot")
 	defer span.End()
 
-	ss, err := st.Flush(ctx)
-	if err != nil {
-		return err
-	}
+	st.snaps.addLayer()
```
Comment (magik6k) [inline chain/state/statetree.go:239]: Why is this needed?

File: chain/state/statetree.go [hunk 5]
```diff
@@ -145,22 +196,32 @@ func (st *StateTree) DeleteActor(addr address.Address) error {
...
+	if err != nil {
+		return err
 	}
 
+	st.snaps.deleteActor(addr)
+
 	return nil
 }
 
 func (st *StateTree) Flush(ctx context.Context) (cid.Cid, error) {
```
Comment (magik6k) [inline chain/state/statetree.go:209]: Can save for a separate PR, but - we could be returning the cache from a flush call, tied to the state root. Then when processing a chain we could be reusing the cache instead of having to rebuild it at each epoch

File: chain/state/statetree.go [hunk 6]
```diff
@@ -26,16 +26,71 @@ type StateTree struct {
...
+		layers: []map[address.Address]streeOp{make(map[address.Address]streeOp)},
+	}
+}
+
+func (ss *stateSnaps) addLayer() {
+	ss.layers = append(ss.layers, make(map[address.Address]streeOp))
+}
+
+func (ss *stateSnaps) dropLayer() {
+	ss.layers = ss.layers[:len(ss.layers)-1]
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply chain/state/statetree.go:53]: Ah, good idea

File: chain/state/statetree.go (hunk 2 above)
Comment (whyrusleeping) [reply chain/state/statetree.go:64]: Yeah, that seems like an optimization we can do later when it comes down to it.

File: chain/state/statetree.go (hunk 4 above)
Comment (whyrusleeping) [reply chain/state/statetree.go:239]: Snapshots add a layer to the snapshot set, that way any changes after we call snapshot go onto their own layer

File: chain/state/statetree.go (hunk 5 above)
Comment (whyrusleeping) [reply chain/state/statetree.go:209]: I like that, that will pair nicely with the next caching thing, which is caching ID lookups

File: chain/state/statetree.go [hunk 7]
```diff
@@ -226,14 +283,9 @@ func (a *AdtStore) Context() context.Context {
...
 
 func (st *StateTree) Revert() error {
-	revTo := st.snapshots[len(st.snapshots)-1]
-	nd, err := hamt.LoadNode(context.Background(), st.Store, revTo, hamt.UseTreeBitWidth(5))
-	if err != nil {
-		return err
-	}
-	st.actorcache = make(map[address.Address]*types.Actor)
+	st.snaps.dropLayer()
+	st.snaps.addLayer()
```
Comment (Kubuxu) [inline chain/state/statetree.go:287]: Why does `Revert` re-add a layer?
